### DISCUSSION TITLE
"Plucking" targets generated by static and dynamic branch for individual consideration

### DISCUSSION THREAD

**Discussion by alexpghayes** [ropensci/targets#419] [Help]
In a paper I am writing, I use `targets` to perform an extensive sensitivity analysis, and I use both static and dynamic branching to fit models across varying specifications. For the most part, I use `tar_combine()` and friends to aggregate across these varying specifications and produce some aggregation results about sensitivity.

However, one of the models that I fit is the model I would actually like to report (i.e. everything else is a sensitivity analysis for that model). I want to do some specific followup work on that model (i.e. nicer plots, more publication ready diagnostics, etc, etc). I believe my problem is equivalent to selecting the target `summary_method2` corresponding to a particular random seed generated by the following (example ripped from https://books.ropensci.org/targets/static.html#dynamic-within-static-branching).

Say I want to apply another function `polish_summary()` to `summary_method2 {random seed i care about}` alone. How would I do that?

```
library(targets)
library(tarchetypes)
library(tibble)
random_seed_target <- tar_target(random_seed, seq_len(10))
targets <- tar_map(
  values = tibble(method_function = rlang::syms(c("method1", "method2"))),
  tar_target(
    analysis,
    method_function("NIH", seed = random_seed),
    pattern = map(random_seed)
  ),
  tar_target(
    summary,
    summarize_analysis(analysis),
    pattern = map(analysis)
  )
)
list(random_seed_target, targets)
```

**Comment by wlandau** [comment]
`targets` has a `slice()` pattern to select an individual branch by its index, but I forgot to document it. As of recent commits to `targets`, `slice()` can take multiple indexes. Sketch:

``` r
library(targets)
tar_script({
  library(targets)
  library(tarchetypes)
  library(tibble)
  
  method1 <- function(label, seed) {
    tibble(method = 1, label = label, data = rnorm(10), seed = seed)
  }
  
  method2 <- function(label, seed) {
    tibble(method = 2, label = label, data = rnorm(10), seed = seed)
  }
  
  summarize_analysis = function(x) {
    tibble(
      method = x$method[1],
      label = x$label[1],
      seed = x$seed[1],
      stat = c("mean", "sd"),
      value = c(mean(x$data), sd(x$data))
    )
  }
  
  targets <- tar_map(
    values = tibble(method_function = rlang::syms(c("method1", "method2"))),
    tar_target(
      analysis,
      method_function("NIH", seed = random_seed),
      pattern = map(random_seed)
    ),
    tar_target(
      summary,
      summarize_analysis(analysis),
      pattern = map(analysis)
    )
  )
  
  list(
    tar_target(random_seed, seq_len(10)),
    targets,
    tar_target(slices, summary_method2, pattern = slice(summary_method2, index = c(3, 4)))
  )
})

tar_make(reporter = "silent")

tar_read(slices)
#> # A tibble: 4 x 5
#>   method label  seed stat   value
#>    <dbl> <chr> <int> <chr>  <dbl>
#> 1      2 NIH       3 mean   0.518
#> 2      2 NIH       3 sd     1.22 
#> 3      2 NIH       4 mean  -0.299
#> 4      2 NIH       4 sd     1.02

tar_read(slices, branches = 1)
#> # A tibble: 2 x 5
#>   method label  seed stat  value
#>    <dbl> <chr> <int> <chr> <dbl>
#> 1      2 NIH       3 mean  0.518
#> 2      2 NIH       3 sd    1.22
```

<sup>Created on 2021-04-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Reply by wlandau** [reply]
Here is another example I just put in the manual:

``` r
library(targets)
tar_script({
  list(
    tar_target(a_data, letters),
    tar_target(b_data, LETTERS),
    tar_target(
      c_data,
      paste(a_data, b_data),
      pattern = slice(cross(a_data, b_data), index = seq(2, 3))
    )
  )
})

tar_make()
#> ● start target a_data
#> ● built target a_data
#> ● start target b_data
#> ● built target b_data
#> ● start branch c_data_5a10da89
#> ● built branch c_data_5a10da89
#> ● start branch c_data_3a92ccbc
#> ● built branch c_data_3a92ccbc
#> ● built pattern c_data
#> ● end pipeline

tar_read(c_data)
#> [1] "a B" "a C"
```

<sup>Created on 2021-04-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Reply by alexpghayes** [reply]
Is there a way to slice by value of `a_data` or `b_data` rather than the index? That would be even nicer

**Reply by wlandau** [reply]
Slicing by value would require more detailed prior knowledge about dynamic branches before they are created, and it would require loading all the upstream branches back into memory to check them. Unfortunately, this would erode the mental model of dynamic branching, break key assumptions in the implementation, and perform inefficiently at scale. So I do not have plans to implement a feature like that.
